src/backbone_router/smcroute_manager.cpp

```diff
@@ -0,0 +1,193 @@
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   The file implements the SMCRoute manager.
+ */
+
+#include "backbone_router/smcroute_manager.hpp"
+
+#include <assert.h>
+#include <unistd.h>
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "common/code_utils.hpp"
+#include "utils/system_utils.hpp"
+
+namespace otbr {
+
+namespace BackboneRouter {
+
+void SMCRouteManager::Init(const std::string &aThreadIfName, const std::string &aBackboneIfName)
+{
+    assert(!mEnabled);
+
+    mThreadIfName   = aThreadIfName;
+    mBackboneIfName = aBackboneIfName;
+
+    StartSMCRouteService();
+}
+
+void SMCRouteManager::Enable(void)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    VerifyOrExit(!mEnabled);
+
+    mEnabled = true;
+
+    Flush();
+
+    // Add mroute rules: 65520 (0xfff0) allow outbound for MA scope >= admin (4)
+    SuccessOrExit(error = AllowOutboundMulticast());
+
+    // Add mroute rules for the current Multicast Listeners Table
+    for (const Ip6Address &address : mListenerSet)
+    {
+        SuccessOrExit(error = AddRoute(address));
+    }
+
+exit:
+    otbrLogResult(error, "SMCRouteManager::Start");
+}
+
+void SMCRouteManager::Disable(void)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    VerifyOrExit(mEnabled);
+
+    mEnabled = false;
+
+    Flush();
+
+    // Remove mroute rules for the current Multicast Listeners Table
+    for (const Ip6Address &address : mListenerSet)
+    {
+        SuccessOrExit(error = DeleteRoute(address));
+    }
+
+    // Remove mroute rules: forbid outbound Multicast traffic
+    SuccessOrExit(error = ForbidOutboundMulticast());
+
+exit:
+    otbrLogResult(error, "SMCRouteManager::Stop");
+}
+
+void SMCRouteManager::StartSMCRouteService(void)
+{
+    otbrError                             error = OTBR_ERROR_NONE;
+    std::chrono::system_clock::time_point deadline;
+
+    VerifyOrExit(system_command("smcroutectl kill || true") == 0, error = OTBR_ERROR_SMCROUTE);
```

Comment:
system_command() is only used by this class, so keep it inside this class?

Reply:
It's can be used by other modules too. As far as we can see, we are going to need to call `ip6tables` and other commands in future Thread 1.2 PRs.